src/Collection/Iterator/UniqueIterator.php

```diff
@@ -0,0 +1,58 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         5.0.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Collection\Iterator;
+
+use Cake\Collection\Collection;
+use Iterator;
+
+/**
+ * Creates a filtered iterator from another iterator. The filtering is done by
+ * passing a callback function to each of the elements and taking them out if
+ * the value returned is not unique.
+ */
+class UniqueIterator extends Collection
+{
+    /**
+     * Creates a filtered iterator using the callback to determine which items are
+     * accepted or rejected.
+     *
+     * Each time the callback is executed it will receive the value of the element
+     * in the current iteration, the key of the element and the passed $items iterator
+     * as arguments, in that order.
+     *
+     * @param iterable $items The items to be filtered.
+     * @param callable $callback Callback.
+     */
+    public function __construct(iterable $items, callable $callback)
+    {
+        if (!$items instanceof Iterator) {
+            $items = new Collection($items);
+        }
+
+        $unique = [];
+        $unique_values = [];
+        foreach ($items as $k => $v) {
+            $compare_value = $callback($v, $k, $items);
```

Comment:
I don't see the purpose of passing the iterator we are currently looping over to the callback. I assume we are trying to mimic CallbackFilterIterator, but can we drop this requirement?